presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java

```diff
@@ -0,0 +1,482 @@
+            pushdownExpressionsBuilder.putAll(remainingProjectExpressions);
+
+            Map<Symbol, Expression> pushdownExpressions = pushdownExpressionsBuilder.build();
+            if (pushdownExpressions.isEmpty()) {
+                return Result.empty();
+            }
+
+            Map<Symbol, Symbol> symbolsMap = pushdownExpressions.entrySet().stream()
+                    .collect(toImmutableMap(entry -> getOnlyElement(extractAll(entry.getValue())), Map.Entry::getKey));
+
+            PlanNode left = joinNode.getLeft();
+            PlanNode right = joinNode.getRight();
+
+            Assignments.Builder leftBuilder = Assignments.builder();
+            leftBuilder.putIdentities(left.getOutputSymbols().stream()
+                    .filter(symbol -> !symbolsMap.containsKey(symbol))
+                    .collect(toImmutableList()));
+
+            Assignments.Builder rightBuilder = Assignments.builder();
+            rightBuilder.putIdentities(right.getOutputSymbols().stream()
+                    .filter(symbol -> !symbolsMap.containsKey(symbol))
+                    .collect(toImmutableList()));
+
+            for (Map.Entry<Symbol, Expression> entry : pushdownExpressions.entrySet()) {
+                Symbol outputSymbol = getOnlyElement(extractAll(entry.getValue()));
+                if (left.getOutputSymbols().contains(outputSymbol)) {
+                    leftBuilder.put(entry.getKey(), entry.getValue());
+                }
+                if (right.getOutputSymbols().contains(outputSymbol)) {
+                    rightBuilder.put(entry.getKey(), entry.getValue());
+                }
+            }
+            ProjectNode leftChild = new ProjectNode(context.getIdAllocator().getNextId(), left, leftBuilder.build());
+            ProjectNode rightChild = new ProjectNode(context.getIdAllocator().getNextId(), right, rightBuilder.build());
+
+            return Result.ofPlanNode(
+                    new JoinNode(
+                        context.getIdAllocator().getNextId(),
+                        joinNode.getType(),
+                        leftChild,
+                        rightChild,
+                        joinNode.getCriteria(),
+                        ImmutableList.<Symbol>builder()
+                                .addAll(leftChild.getOutputSymbols())
+                                .addAll(rightChild.getOutputSymbols())
+                                .build(),
+                        joinNode.getFilter().map(expression -> replaceDereferences(expression, dereferenceSymbols)),
+                        joinNode.getLeftHashSymbol(),
+                        joinNode.getRightHashSymbol(),
+                        joinNode.getDistributionType()));
+        }
+    }
+
+    /** Transforms:
+     * <pre>
+     *  Project(a_x := a.msg.x)
+     *    Project(a_y := key)
+     *          Source(a)
+     *  </pre>
+     * to:
+     * <pre>
+     *  Project(a_x := a_z)
```

Comment:
This project node is redundant. Why not remove it?

Reply:
there is RemoveRedundantIdentityProjections following PushDownDereferences, so that redundant project node is removed. I will remove it from comments. Or, do you think we should add RemoveRedundantIdentityProjections to PushDownDereferences?